File: app/src/main/java/com/example/leo/test/MediaFragment.java
package com.example.leo.test;

import android.arch.lifecycle.ViewModelProviders;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.databinding.DataBindingUtil;
import android.databinding.ObservableField;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.content.LocalBroadcastManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.SeekBar;
import android.widget.TextView;

import com.example.leo.test.databinding.FragmentMediaBinding;
import java.util.Locale;

import static android.content.Context.BIND_AUTO_CREATE;
import static com.example.leo.test.MediaPlayerService.MEDIA_PATH_KEY;

public class MediaFragment extends Fragment {

    private MediaPlayerService mMediaPlayerService;
    private final ObservableField<Boolean> mIsPlaying = new ObservableField<>(true);

    private Button mPlayButton;
    private Button mNextButton;
    private Button mLastButton;
    private SeekBar mSeekBar;
    private TextView mCurrentProgress;
    private TextView mTotalLength;

    private LocalBroadcastManager mLocalBroadcastManager;

    private MediaListViewModel mMediaListViewModel;

    private ServiceConnection mMediaPlayerConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            mMediaPlayerService = ((MediaPlayerService.MediaPlayerBinder) service).getService();

            mLocalBroadcastManager.registerReceiver(mMediaReadyBroadcastReceiver, new IntentFilter(MediaPlayerService.MEDIA_READY));

            Intent playIntent = new Intent(getContext(), MediaPlayerService.class);
            playIntent.putExtra(MEDIA_PATH_KEY, getArguments().getString(MediaPlayerService.MEDIA_PATH_KEY));
            getContext().startService(playIntent);
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            mMediaPlayerService = null;
        }
    };

    private BroadcastReceiver mMediaReadyBroadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            setupViewStatus();
            mPlayButton.setEnabled(true);

            mHandler.post(mUpdateUI);
        }
    };

    private Handler mHandler = new Handler((Looper.getMainLooper()));
    private Runnable mUpdateUI = new Runnable() {
        @Override
        public void run() {
            final int progress = mMediaPlayerService.getProgress();
            mSeekBar.setProgress(progress);
            mCurrentProgress.setText(formatMilliSecLength(progress));
            mHandler.postDelayed(mUpdateUI, 1000);
        }
    };

    public static MediaFragment newInstance(String mediaPath) {
        MediaFragment fragment = new MediaFragment();

        Bundle args = new Bundle();
        args.putString(MediaPlayerService.MEDIA_PATH_KEY, mediaPath);
        fragment.setArguments(args);

        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        setRetainInstance(true);

        FragmentMediaBinding binding = DataBindingUtil.inflate(
                inflater, R.layout.fragment_media, container, false);
        setupDataBinding(binding);

        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        view.setBackgroundColor(Color.WHITE);

        mPlayButton = view.findViewById(R.id.playButton);
        mLastButton = view.findViewById(R.id.lastMediaButton);
        mNextButton = view.findViewById(R.id.nextMediaButton);

        mSeekBar = view.findViewById(R.id.progressSeekBar);
        mCurrentProgress = view.findViewById(R.id.currentProgress);
        mTotalLength = view.findViewById(R.id.totalLength);

        mLocalBroadcastManager = LocalBroadcastManager.getInstance(getContext());

        mMediaListViewModel = ViewModelProviders.of(getActivity()).get(MediaListViewModel.class);
    }

    @Override
    public void onStart() {
        super.onStart();
        getContext().getApplicationContext().bindService(new Intent(getContext(), MediaPlayerService.class), mMediaPlayerConnection, BIND_AUTO_CREATE);
    }

    private void setupViewStatus() {
        mSeekBar.setMax(mMediaPlayerService.getMediaLength());
        mSeekBar.setProgress(mMediaPlayerService.getProgress());
        mTotalLength.setText(formatMilliSecLength(mMediaPlayerService.getMediaLength()));
        mCurrentProgress.setText(formatMilliSecLength(mMediaPlayerService.getProgress()));

        mSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (fromUser) {
                    mMediaPlayerService.seekToPosition(mSeekBar.getProgress());
                    mCurrentProgress.setText(formatMilliSecLength(mMediaPlayerService.getProgress()));
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });
    }

    private String formatMilliSecLength(int millis) {
        final int sec = millis / 1000;

        return String.format(Locale.US, "%02d:%02d", sec / 60, sec % 60);
    }

    private void setupDataBinding(FragmentMediaBinding binding) {
        binding.setMainView(this);
        binding.setIsPlaying(mIsPlaying);
    }

    @Override
    public void onStop() {
        super.onStop();
        // Remove scheduled access to the service
        mHandler.removeCallbacks(mUpdateUI);

        mLocalBroadcastManager.unregisterReceiver(mMediaReadyBroadcastReceiver);

        mMediaPlayerService = null;
        getContext().getApplicationContext().unbindService(mMediaPlayerConnection);
    }

    public void playOrPauseMediaPlay(View view) {
        assert (mIsPlaying.get() != null);

        if (mIsPlaying.get()) {
            mMediaPlayerService.pause();
        } else {
            mMediaPlayerService.play();
        }

        mIsPlaying.set(!mIsPlaying.get());
    }

    public void previousMedia(View view) {
        mMediaPlayerService.resetTo(mMediaListViewModel.moveToPreviousMedia());
    }

    public void nextMedia(View view) {
        mMediaPlayerService.resetTo(mMediaListViewModel.moveToNextMedia());
    }
}

File: app/src/main/kotlin/com/example/leo/test/MediaFragment.kt
package com.example.leo.test

import android.content.*
import android.content.Context.BIND_AUTO_CREATE
import android.databinding.DataBindingUtil
import android.databinding.ObservableField
import android.graphics.Color
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.support.v4.app.Fragment
import android.support.v4.content.LocalBroadcastManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.SeekBar
import com.example.leo.test.MediaPlayerService.Companion.MEDIA_PATH_KEY
import com.example.leo.test.databinding.FragmentMediaBinding
import java.util.*
import kotlinx.android.synthetic.main.fragment_media.*

class MediaFragment : Fragment() {

    private var mediaPlayerService: IMediaPlayerService? = null
    private val isPlaying = ObservableField(true)
    private val title = ObservableField("Title")
    private val author = ObservableField("Author")

    private lateinit var localBroadcastManager: LocalBroadcastManager

    private val mediaPlayerConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            mediaPlayerService = (service as MediaPlayerService.MediaPlayerBinder).service
        }

        override fun onServiceDisconnected(name: ComponentName) {
            mediaPlayerService = null
        }
    }

    private val handler = Handler(Looper.getMainLooper())

    private val mediaReadyBroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            setupViewStatus()
            playButton.isEnabled = true
        }
    }

    private val mediaChangedBroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            setupViewStatus()
        }
    }

    private val updateUI = object : Runnable {
        override fun run() {
            mediaPlayerService?.progress?.let {
                progressSeekBar.progress = it
                currentProgress.text = formatMilliSecLength(it)
            }
            handler.postDelayed(this, 1000)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        retainInstance = true

        val binding: FragmentMediaBinding = DataBindingUtil.inflate(
                inflater, R.layout.fragment_media, container, false)
        setupDataBinding(binding)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        view.setBackgroundColor(Color.WHITE)

        activity?.let {
            localBroadcastManager = LocalBroadcastManager.getInstance(it).apply {
                registerReceiver(mediaReadyBroadcastReceiver, IntentFilter(MediaPlayerService.MEDIA_READY))
                registerReceiver(mediaChangedBroadcastReceiver, IntentFilter(MediaPlayerService.NEXT_MEDIA))
                registerReceiver(mediaChangedBroadcastReceiver, IntentFilter(MediaPlayerService.PREVIOUS_MEDIA))
            }
        }

        val playIntent = Intent(context, MediaPlayerService::class.java)
        playIntent.putExtra(MEDIA_PATH_KEY, arguments?.getString(MediaPlayerService.MEDIA_PATH_KEY))
        context?.startService(playIntent)
    }

    override fun onStart() {
        super.onStart()

        handler.post(updateUI)

        context?.bindService(
                Intent(context, MediaPlayerService::class.java),
                mediaPlayerConnection, BIND_AUTO_CREATE)
    }

    override fun onResume() {
        super.onResume()
        setupViewStatus()
    }

    override fun onStop() {
        super.onStop()
        // Remove scheduled access to the service
        handler.removeCallbacks(updateUI)

        localBroadcastManager.apply {
            unregisterReceiver(mediaReadyBroadcastReceiver)
            unregisterReceiver(mediaChangedBroadcastReceiver)
        }

        mediaPlayerService = null
        context?.unbindService(mediaPlayerConnection)
    }

    private fun setupViewStatus() {
        mediaPlayerService?.run {
            currentMedia?.let {
                title.set(it.title)
                author.set(it.artist)
            }

            progressSeekBar.max = mediaLength
            progressSeekBar.progress = progress
            totalLength.text = formatMilliSecLength(mediaLength)
            currentProgress.text = formatMilliSecLength(progress)

            progressSeekBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
                override fun onProgressChanged(progressSeekBar: SeekBar, progress: Int, fromUser: Boolean) {
                    if (fromUser) {
                        seekToPosition(this@MediaFragment.progressSeekBar.progress)
                        currentProgress.text = formatMilliSecLength(progress)
                    }
                }

                override fun onStartTrackingTouch(progressSeekBar: SeekBar) {

                }

                override fun onStopTrackingTouch(progressSeekBar: SeekBar) {

                }
            })
        }
    }

    private fun formatMilliSecLength(millis: Int): String {
        val sec = millis / 1000

        return String.format(Locale.US, "%02d:%02d", sec / 60, sec % 60)
    }

    private fun setupDataBinding(binding: FragmentMediaBinding) {
        binding.mainView = this
        binding.isPlaying = isPlaying
        binding.title = title
        binding.author = author
    }

    fun playOrPauseMediaPlay(view: View) {
        isPlaying.run {
            if (get() == true) {
                mediaPlayerService?.pause()
            } else {
                mediaPlayerService?.play()
            }

            set(get()?.not())
        }
    }

    fun previousMedia(view: View) {
        mediaPlayerService?.previousMedia()
    }

    fun nextMedia(view: View) {
        mediaPlayerService?.nextMedia()
    }

    companion object {
        fun newInstance(mediaPath: String) = MediaFragment().apply {
            arguments = Bundle().apply {
                putString(MediaPlayerService.MEDIA_PATH_KEY, mediaPath)
            }
        }
    }
}
